src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Linux.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Sockets
+{
+    sealed class SocketAsyncEngine
+    {
+        private static SocketAsyncEngine _engine;
+        private static readonly object _initLock = new object();
+
+        private readonly int _epollFd;
+
+        public static SocketAsyncEngine Instance
+        {
+            get
+            {
+                if (Volatile.Read(ref _engine) == null)
+                {
+                    lock (_initLock)
+                    {
+                        if (_engine == null)
+                        {
+                            int epollFd = Interop.libc.epoll_create1(Interop.libc.EPOLL_CLOEXEC);
+                            if (epollFd == -1)
+                            {
+                                // TODO: throw an appropriate exception
+                                throw new InternalException();
+                            }
+
+                            var engine = new SocketAsyncEngine(epollFd);
+                            Task.Factory.StartNew(o => {
+                                ((SocketAsyncEngine)o).EventLoop();
+                            }, engine, TaskCreationOptions.LongRunning);
```

Comment:
I'd suggest explicitly adding TaskScheduler.Default here as the scheduler argument.  Two related reasons:
- You want to make sure this actually targets the thread pool scheduler even if it happens to be called from a task running on another scheduler.
- The behavior of LongRunning is controlled by the scheduler: the thread pool scheduler gives it a dedicated thread, but another scheduler could behave differently.